crates/dbsp/src/trace/cursor/mod.rs

```diff
@@ -306,6 +308,52 @@ pub trait Cursor<K: ?Sized, V: ?Sized, T, R: ?Sized> {
     }
 }
 
+/// An object that can produce a cursor bounded by its lifetime.
+///
+/// [BatchReader::cursor] produces a cursor bounded by the [BatchReader]'s
+/// lifetime.  Suppose [BatchReader::cursor] could use an existing [Cursor]
+/// implementation, such as [CursorList], but it needs to create some data for
+/// the cursor to own.  Then it's rather inconvenient because it's necessary to
+/// create a new type to own the data and implement the whole broad [Cursor]
+/// trait to forward every call to [CursorList]. Plus, you end up with a
+/// self-referencing type, so you need to use [ouroboros].  See `SpineCursor` in
+/// the async spine for a good example.
+///
+/// Suppose we're building a new interface where we want to return a `dyn
+/// Cursor`, and some of the implementations would suffer from the problem
+/// above.  We can instead return a [CursorFactory].  An implementation of this
+/// trait can own the data that it needs to, and then implement
+/// [CursorFactory::get_cursor] to create and return a cursor whose lifetime is
+/// bounded by the [CursorFactory].  This reduces the boilerplate a great deal
+/// (and we don't need [ouroboros], either).
```

Comment:
from a very superficial point of view this looks like something where GAT could help

Reply:
Good point. Do you know whether GATs are object-safe now? They initially were not: https://blog.rust-lang.org/2022/10/28/gats-stabilization/#traits-with-gats-are-not-object-safe. This bug suggests that they still are not: https://github.com/rust-lang/rust/issues/81823.